### ISSUE TITLE
Resource patch fails if `mergeStrategy: merge` or `strategic` and patch file with `flags` provided

### ISSUE THREAD

**Issue by vanny96** [argoproj/argo-workflows#11248]
### Pre-requisites

- [X] I have double-checked my configuration
- [X] I can confirm the issues exists when I tested with `:latest`
- [X] I'd like to contribute the fix myself (see [contributing guide](https://github.com/argoproj/argo-workflows/blob/master/docs/CONTRIBUTING.md))

### What happened/what you expected to happen?

Whenever a "resource patch" is performed with a `mergeStrategy` different from `json` the patching fails.
This error seems to have been introduced with the [this PR](https://github.com/argoproj/argo-workflows/pull/5951)

My recommended solution would be to revert this snippet of code in `workflow/executor/resource.go`
```go
	appendFileFlag := true
	if action == "patch" {
		mergeStrategy := "strategic"
		if we.Template.Resource.MergeStrategy != "" {
			mergeStrategy = we.Template.Resource.MergeStrategy
			if mergeStrategy == "json" {
				// Action "patch" require flag "-p" with resource arguments.
				// But kubectl disallow specify both "-f" flag and resource arguments.
				// Flag "-f" should be excluded for action "patch" here if it's a json patch.
				appendFileFlag = false
			}
		}

		args = append(args, "--type")
		args = append(args, mergeStrategy)
		args = append(args, "-p")
		args = append(args, string(buff))
	}
	if len(flags) != 0 {
		args = append(args, flags...)
	}

	if len(buff) != 0 && appendFileFlag {
		args = append(args, "-f")
		args = append(args, manifestPath)
	}
```

to
```go
	if action == "patch" {
		mergeStrategy := "strategic"
		if we.Template.Resource.MergeStrategy != "" {
			mergeStrategy = we.Template.Resource.MergeStrategy
		}

		args = append(args, "--type")
		args = append(args, mergeStrategy)
		args = append(args, "-p")
		args = append(args, string(buff))
	}
	if len(flags) != 0 {
		args = append(args, flags...)
	}

	// Action "patch" require flag "-p" with resource arguments.
	// But kubectl disallow specify both "-f" flag and resource arguments.
	// Flag "-f" should be excluded for action "patch" here.
	if len(buff) != 0 && action != "patch" {
		args = append(args, "-f")
		args = append(args, manifestPath)
	}
```

### Version

latest

### Paste a small workflow that reproduces the issue. We must be able to run the workflow; don't enter a workflows that uses private images.

```YAML
metadata:
  name: test-workflow
spec:
  entrypoint: main
  templates:
    - name: main
      steps:
        - - name: create-cm
            template: create-cm

        - - name: patch-cm
            template: patch-cm
            arguments:
              parameters:
                - name: config-map-name
                  value: {{ steps.create-cm.outputs.config-map-name }}

    - name: create-cm
      resource:
        action: create
        manifest: |
          apiVersion: v1
          kind: ConfigMap
          metadata:
            generateName: config-map-test-
          data:
            test: "Hello World"
      outputs:
        parameters:
          - name: config-map-name
            valueFrom:
              jsonPath: '{.metadata.name}'

    - name: patch-cm
      inputs:
        parameters:
          - name: config-map-name
      resource:
        action: patch
        mergeStrategy: merge
        flags:
          - cm
          - "{{ inputs.parameters.config-map-name }}"
        manifest: |
          {
            "data": {
              "test": "Goodbye"
            }
          }
```

### Logs from the workflow controller

```text
time="2023-06-21T09:29:13 UTC" level=info msg="kubectl patch --type merge -p {\n  \"data\": {\n\"test\": \"Goodbye\"\n  }\n}\n cm -f /tmp/manifest.yaml -o json"
time="2023-06-21T09:29:13 UTC" level=warning msg="Non-transient error: exit status 1"
time="2023-06-21T09:29:13 UTC" level=error msg="executor error: no more retries error: when paths, URLs, or stdin is provided as input, you may not specify resource arguments as well"
time="2023-06-21T09:29:13 UTC" level=fatal msg="no more retries error: when paths, URLs, or stdin is provided as input, you may not specify resource arguments as well"
```

### Logs from in your workflow's wait container

```text
Nothing of relevance
```

**Comment by vanny96** [comment]
I'll create a PR shortly

**Comment by agilgur5** [comment]
This was closed per https://github.com/argoproj/argo-workflows/pull/11249#issuecomment-1600745498

**Comment by agilgur5** [comment]
> This was closed per [#11249 (comment)](https://github.com/argoproj/argo-workflows/pull/11249#issuecomment-1600745498)

Turns out that is actually a workaround and this bug is accurate in a few cases, as discussed in #13279

I've fixed that bug in #13317 now and provided more examples and tests there as well.
